pkg/logs/client/destination.go

```diff
@@ -74,3 +85,35 @@ func (d *Destination) Send(payload []byte) error {
 
 	return nil
 }
+
+// SendAsync sends a message to the destination without blocking. If the queue is full, the incoming messages will be
+// dropped
+func (d *Destination) SendAsync(payload []byte) {
+	d.once.Do(func() {
+		inputChan := make(chan []byte, chanSize)
+		d.inputChan = inputChan
+		go d.consumeAsync()
+	})
+
+	select {
+	case d.inputChan <- payload:
+	default:
+		// TODO: Display the warning in the status
+		if d.warningCounter%warningPeriod == 0 {
+			log.Warnf("Some logs sent to additional destination %v were dropped", d.connManager.endpoint.Host)
+		}
+		d.warningCounter++
+	}
+}
+
+// consumeAsync read the messages from the queue and send them
+func (d *Destination) consumeAsync() {
+	ctx := d.destinationsContext.Context()
+	for {
+		select {
+		case payload := <-d.inputChan:
+			d.Send(payload)
```

Comment:
This call is still blocking, how do you drop the message from here ? Shall we add a timeout to the send method ?

Reply:
The timeout/retries is handled by the `ConnectionManager` 
It's not a problem to be blocking here because we are on the consuming end of the additional destinations. It is not in the main destination sending loop

Reply:
What we have now is:
- one message blocking
- all the next messages dropped
Would a fire and forget solution be better ? i.e. try once to send a message so that every message has a chance to be sent when the network connection is slow ? This would imply to break some logic though

Reply:
WDYT ?

Reply:
Yes, I agree that it would be better to drop older logs before newer one.

I'm not sure about how we could implement this though:

- In the case of a wrong configuration, the `ConnectionManager.NewConnection()` call will be blocking, and it doesn't really matter because all the messages will be dropped
- In the case of a a problem that causes the additional destination to be unreachable for a limited amount of time (e.g. network partition):
  - Either the  `ConnectionManager.NewConnection()` call will be blocking
  - Or the `d.conn.Write(frame)` call will be blocking 
- If the connection to the `AdditionalDestination` is slower than the `MainDestination`, the `SendAsync` channel will slowly be filled then drop some logs

We might want to timeout mechanism for the blocking calls, but if we timeout after 1 sec and we send 1000 logs / s, we will still drop the 900 most recent logs, then the oldest one, then 999 most recent logs, then the 2nd oldest one, etc. It's a weird behaviour.

I don't see how we can have a better behaviour with `SendAsync` if we have a fixed size buffer or without running the `Send` in a goroutine

Reply:
For additional destinations, it is best effort, so let's do as simple as we can